Toolsmiths/add build tags for special builds

tdctaz opened the pull request:
The purpose of this is to make it more visible that a build has been executed under some special cases, for example split into multiple partitions, has only partial run tests or has some filter enabled. Anything to indicate that the build is not normal.

The configuration of the build tags can be calculated from the build properties allowing it to be dynamic and be applied to past builds as well.

Example of a configuration setup for build tags that tags the builds with even or odd tags dependent on the build number looks something like this:

```
def EvenOrOddTag(properties):
    if properties.hasProperty('buildnumber'):
        buildnumber = int(properties.getProperty('buildnumber'))
        if buildnumber % 2 == 0:
            return BuildTag('Even', 'The build with the number %d is even' % buildnumber)
        else:
            return BuildTag('Odd', 'The build with the number %d is odd' % buildnumber)

c['builders'].append(
    BuilderConfig(name="Simple Config",
        project="Sample",
        build_tags=[
            EvenOrOddTag,
            BuildTag('Foobar'),
        ],
        slavenames=hg_slaves,
        factory=simpleConfigFactory())
)
```
The frontend will then render the build tags on the builds them self under the build number of the build status or on the main build page, see the examples below:

https://i.imgur.com/dgDwR15.png
https://i.imgur.com/wd9ZCAV.png
https://i.imgur.com/Afc5oVF.png

unity-christianb committed:
Update production javascript

unity-christianb committed:
Add build tags to the build page

unity-christianb committed:
Add BuildTags to Katana

unity-christianb committed:
Refactor: indentation fix

MindWrapper commented on master/buildbot/config.py:
```diff
@@ -816,6 +816,7 @@ def __init__(self, name=None, slavename=None, startSlavenames=None, slavenames=N
         self.mergeRequests = mergeRequests
         self.project = project
         self.tags = tags
+        self.build_tags = build_tags
```
minor: variable name style mismatch

tdctaz replied:
yes, but the style is 50/50 in the code base so it is hard to choose :)

tdctaz replied:
kept it as is, since `builder_tags` is written with `_` and this is closely related - when the code style is chaotic then in a very chaotic way this somehow makes sense to me :)

mariangemarcano replied:
@tdctaz @MindWrapper this is the reason why it is what it is.
Twisted prefers !camelCase for variables, functions, and methods (whereas PEP 8 recommends underscores) 
http://twistedmatrix.com/trac/wiki/FrequentlyAskedQuestions#WhydoesntTwistedfollowPEP8 
In general, we follow Twisted camelCase style. But given that the project is in maintenance mode there is no plan for us to fix the code style issues.

mariangemarcano commented on master/buildbot/status/build.py:
```diff
@@ -629,6 +631,23 @@ def get_artifacts(self):
 
+    def getBuildTags(self):
+        build_tags = []
+        builder = self.master.botmaster.builders[self.builder.name]
```
@tdctaz can you change it to self.botmaster.builders.get instead

unity-christianb committed:
Fix use get instead of dictionary lookup

unity-christianb committed:
Add tests for buildtags

unity-christianb committed:
Fix imports

unity-christianb committed:
Fix wrong builder path

Since the test now setup it's buildbot.builders collection properly the generated path to the build is now also correct. See `def getBuildersPath` in master.py

tdctaz commented:
@mariangemarcano The test has been added and fixed - after I hooked up the botmaster.builders for the test the url changed for some of the tests to the correct url.
